navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt

```diff
@@ -165,63 +177,78 @@ public class BottomSheetNavigator(
-        // window destinations. That means that only *one* concurrent destination is supported by
-        // this navigator.
-        val latestEntry = backStackEntries.lastOrNull { entry ->
-            // We might have entries in the back stack that aren't started currently, so filter
-            // these
-            entry.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)
-        }
+    val sheetContent: @Composable ColumnScope.() -> Unit =
+        movableContentWithReceiverOf<ColumnScope> {
+            val saveableStateHolder = rememberSaveableStateHolder()
+            val transitionsInProgressEntries by transitionsInProgress.collectAsState()
 
-        // Mark all of the entries' transitions as complete, except for the entry we are
-        // currently displaying because it will have its transition completed when the sheet's
-        // animation has completed
-        DisposableEffect(backStackEntries) {
-            transitionsInProgressEntries.forEach {
-                if (it != latestEntry) state.markTransitionComplete(it)
+            val retainedEntry by produceState<NavBackStackEntry?>(initialValue = null) {
+                backStack
+                    .map { backStackEntries ->
+                        backStackEntries.lastOrNull { entry ->
+                            // We might have entries in the back stack that aren't started
+                            // currently, so filter these
+                            entry.lifecycle.currentState.isAtLeast(Lifecycle.State.INITIALIZED)
+                        }
+                    }
+                    //.distinctUntilChanged()
+                    .transform { latestEntry ->
+                        sheetState.hide()
+                        emit(latestEntry)
+                    }
+                    .collectLatest { value = it }
             }
-            onDispose { }
-        }
 
-        SheetContentHost(
-            columnHost = columnScope,
-            backStackEntry = latestEntry,
-            sheetState = sheetState,
-            saveableStateHolder = saveableStateHolder,
-            onSheetShown = { backStackEntry ->
-                state.markTransitionComplete(backStackEntry)
-            },
-            onSheetDismissed = { backStackEntry ->
-                // Sheet dismissal can be started through popBackStack in which case we have a
-                // transition that we'll want to complete
-                if (transitionsInProgressEntries.contains(backStackEntry)) {
-                    state.markTransitionComplete(backStackEntry)
-                } else {
-                    state.pop(popUpTo = backStackEntry, saveState = false)
+            // Mark all of the entries' transitions as complete, except for the entry we are
+            // currently displaying because it will have its transition completed when the sheet's
+            // animation has completed
+            DisposableEffect(transitionsInProgressEntries) {
```

Comment:
Can we still change the system so that we always do the show/hide so that we get callbacks in all cases. And then the `onSheetShown` and `onSheetDismissed` callbacks can just do:

```
transitionsInProgressEntries.forEach {
    state.markTransitionComplete(it)
}
```

And this big of code should no longer be needed I believe.

Reply:
Done